File: 11505.cpp
#include<stdio.h>
#include<vector>
#include<cmath>

#define left(a) a*2+1
#define right(a) a*2+2
#define MOD 1000000007

using namespace std;

int numOfNumbers, numOfUpdates, numOfOperations;
vector<long long> numbers = {0};
vector<long long> segmentTree;

void input();
long long initTree(int from, int to, int node);
void solve();
long long update(int target, long long newValue, int from, int to, int node);
long long mul(int rangeFrom, int rangeTo, int from, int to, int node);

int main() {
    input();
    segmentTree = vector<long long>((1 << int(ceil(log2(numOfNumbers)) + 1)) - 1, 0);
    initTree(1, numOfNumbers, 0);
    solve();
    return 0;
}

void input() {
    scanf("%d %d %d", &numOfNumbers, &numOfUpdates, &numOfOperations);
    
    for(int i = 0; i < numOfNumbers; i++) {
        long long num;
        scanf("%lld", &num);
        numbers.push_back(num);
    }
}

long long initTree(int from, int to, int node) {
    if(from == to) {
        segmentTree[node] = numbers[from];
    } else {
        int mid = (from + to) / 2;
        long long left = initTree(from, mid, left(node));
        long long right = initTree(mid + 1, to, right(node));
        segmentTree[node] = (left * right) % MOD;
    }
    return segmentTree[node];
}

void solve() {
    for(int i = 0; i < numOfUpdates + numOfOperations; i++) {
        int command, a;
        long long b;
        scanf("%d %d %lld", &command, &a, &b);
        
        if(command == 1) {
            update(a, b, 1, numOfNumbers, 0);
        } else if(command == 2) {
            printf("%lld\n", mul(a, int(b), 1, numOfNumbers, 0));
        }
    }
}

long long mul(int rangeFrom, int rangeTo, int from, int to, int node) {
    if(from > rangeTo || to < rangeFrom) {
        return 1;
    } else if(rangeFrom <= from && to <= rangeTo) {
        return segmentTree[node];
    } else {
        int mid = (from + to) / 2;
        long long left = mul(rangeFrom, rangeTo, from, mid, left(node));
        long long right = mul(rangeFrom, rangeTo, mid + 1, to, right(node));
        return (left * right) % MOD;
    }
}

long long update(int target, long long newValue, int from, int to, int node) {
    if(from == to) {
        if(from == target) {
            segmentTree[node] = newValue;
            numbers[target] = newValue;
        }
        return segmentTree[node];
    } else if(from <= target && target <= to) {
        int mid = (from + to) / 2;
        long long left = update(target, newValue, from, mid, left(node));
        long long right = update(target, newValue, mid + 1, to, right(node));
        segmentTree[node] = (left * right) % MOD;
        return segmentTree[node];
    } else {
        return segmentTree[node];
    }
}

File: 11505.swift
import Foundation

let operand: Int64 = 1_000_000_007

let input = readLine()!.split(separator: " ").map({Int(String($0))!})
let numOfNumbers = input[0]
let numOfCommands = input[1] + input[2]

var numbers: [Int64] = [0]

for _ in 0..<numOfNumbers {
    numbers.append(Int64(readLine()!)!)
}

var segmentTree: [Int64] = Array(repeating: 0, count: (1 << Int(ceil(log2(Double(numOfNumbers))) + 1)) - 1)
let _ = buildTree(from: 1, to: numOfNumbers, node: 0)

for _ in 0..<numOfCommands {
    let command = readLine()!.split(separator: " ").map({Int64(String($0))!})
    
    if command[0] == 1 {
        let target = Int(command[1])
        let newValue = command[2]
        let _ = update(target: target, newValue: newValue, from: 1, to: numOfNumbers, node: 0)
    } else if command[0] == 2 {
        let rangeFrom = Int(command[1])
        let rangeTo = Int(command[2])
        print(mul(rangeFrom: rangeFrom, rangeTo: rangeTo, from: 1, to: numOfNumbers, node: 0))
    }
}


func leftChild(of node: Int) -> Int {
    return node * 2 + 1
}

func rightChild(of node: Int) -> Int {
    return node * 2 + 2
}

func buildTree(from: Int, to: Int, node: Int) -> Int64 {
    if from == to {
        segmentTree[node] = numbers[from]
    } else {
        let mid = (from + to) / 2
        let left = buildTree(from: from, to: mid, node: leftChild(of: node))
        let right = buildTree(from: mid + 1, to: to, node: rightChild(of: node))
        segmentTree[node] = (left * right) % operand
    }
    return segmentTree[node]
}

func mul(rangeFrom: Int, rangeTo: Int, from: Int, to: Int, node: Int) -> Int64 {
    if rangeFrom > to || rangeTo < from {
        return 1
    } else if rangeFrom <= from && to <= rangeTo {
        return segmentTree[node]
    } else {
        let mid = (from + to) / 2
        let left = mul(rangeFrom: rangeFrom, rangeTo: rangeTo, from: from, to: mid, node: leftChild(of: node))
        let right = mul(rangeFrom: rangeFrom, rangeTo: rangeTo, from: mid + 1, to: to, node: rightChild(of: node))
        return (left * right) % operand
    }
}

func update(target: Int, newValue: Int64, from: Int, to: Int, node: Int) -> Int64 {
    if from == to {
        if from == target {
            segmentTree[node] = newValue
            numbers[from] = newValue
        }
        return segmentTree[node]
    } else if from <= target && target <= to {
        let mid = (from + to) / 2
        let left = update(target: target, newValue: newValue, from: from, to: mid, node: leftChild(of: node))
        let right = update(target: target, newValue: newValue, from: mid + 1, to: to, node: rightChild(of: node))
        segmentTree[node] = (left * right) % operand
    }
    return segmentTree[node]
}

